src/handler/handler.rs

```diff
@@ -84,3 +90,74 @@ pub async fn signin<'a>(
         }
     }
 }
+
+#[get("/api/sessions/oauth/google?<code>&<state>")]
+pub async fn google_oauth(
+    code: String,
+    state: String,
+    conn: DbConn,
+    // jar: &CookieJar<'_>,
+) -> Result<Redirect, (Status, String)> {
+    if code.is_empty() {
+        return Err((
+            Status::Unauthorized,
+            json!({"status": "fail", "message": "Authorization code not provided!"}).to_string(),
+        ));
+    }
+
+    let token_response = request_token(&code).await;
+    if let Err(e) = token_response {
+        return Err((
+            Status::BadGateway,
+            json!({"status": "fail", "message": e.to_string()}).to_string(),
+        ));
+    }
+
+    let token_response = token_response.unwrap();
+    let google_user = get_google_user(&token_response.access_token, &token_response.id_token).await;
+    if let Err(e) = google_user {
+        return Err((
+            Status::BadGateway,
+            json!({"status": "fail", "message": e.to_string()}).to_string(),
+        ));
+    }
+
+    let google_user = google_user.unwrap();
+    let user_email = google_user.email;
+
+    let result = conn
+        .run(move |c| {
+            // We wanted to verify if in database it is existed or not
+            // If not then insert in database else just get user info
+
+            let user = users
+                .filter(email.eq(user_email))
+                .first::<User>(c)
+                .optional();
+
+            // User not existed
+            if let Ok(Some(_user)) = &user {
+                // Insert user to db as well
+            }
+
+            return user;
+        })
+        .await;
+
+    // Making of JWT token
+    let _: Json<Result<String, String>> = match result {
+        Ok(Some(user)) => {
+            let token = generate_token(&user.username)
+                .unwrap_or_else(|_| "Failed to generate token".to_string());
+            Json(Ok(token))
+        }
+        Ok(None) => Json(Err("Invalid credentials".into())),
+        Err(err) => {
+            eprintln!("Error during sign in: {:?}", err);
+            Json(Err("Error during sign in".into()))
+        }
+    };
+
+    let frontend_origin = std::env::var("CLIENT_ORIGIN").unwrap();
+    Ok(Redirect::to(format!("{}{}", frontend_origin, state)))
```

Comment:
1. Do we have to redirect the user to any particular url?
2. Should we generate JWT token & return? or just set in cookies jar